File: chapter02/denotationalSemantic/If.rb
class If < Struct.new(:condition, :consequence, :alternative)
  def to_ruby
    "-> e { if (#{condition.to_ruby}).call(e)" +
        " then (#{consequence.to_ruby}).call(e)" +
        " else (#{alternative.to_ruby}).call(e)" +
        "end }"
  end
end

File: chapter02/DenotationSemantic/If.py
class If(object):
    def __init__(self, condition, consequence, alternative):
        self.condition = condition
        self.consequence = consequence
        self.alternative = alternative

    def to_python(self):
        condition = self.condition.to_python()
        consequence = self.consequence.to_python()
        alternative = self.alternative.to_python()
        return 'if {condition} :\n    {consequence}\nelse:\n    {alternative}'.format(**locals())